lib/src/util/useDynamicDropdownPlacement.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import { useState, useEffect, RefObject } from 'react';
+
+/**
+ * Convenience hook that determines whether a dropdown menu should open on the top or on the right
+ * instead of opening the menu on the bottom or on the left respectively.
+ * This custom hook basically checks the position of the dropdown based on where it is in the current page,
+ * and determines if the dropdown menu should be placed on the top or on the right based on scroll or resize.
+ * @param dropdownDivRef the ref to the div that holds the dropdown menu.
+ * @param childrenCount the number of items the dropdown menu contains to dynamically allocate spacing.
+ * @return a [boolean, boolean] if the dropdown menu should be placed on top and on the right respectively.
+ */
+
+const DROPDOWN_ITEM_HEIGHT = 32; //32px is the height of an item in the dropdown
+const ICON_ONLY_DROPDOWN_WIDTH = 250; //250px is the width of the dropdown
+
+const useDynamicDropdownPlacement = (
+  dropdownDivRef: RefObject<HTMLDivElement>,
+  childrenCount: number,
+) => {
+  const [openTop, setOpenTop] = useState<boolean>(false);
+  const [openRight, setOpenRight] = useState<boolean>(false);
+
+  // The event listeners look for changes to be made to the dropdown position if the user
+  // changes the window size by either scrolling the page or resizing the page.
+  // childrenCount is added as a dependency because in components like NxSearchDropdown, the height of the
+  // menu depends on the number of results/menu items shown in the component.
+  useEffect(() => {
+    window.addEventListener('scroll', determinePlacement);
+    window.addEventListener('resize', determinePlacement);
```

Comment:
Consider `useResizeObserver` to check for resizes of the element's container, not just the viewport